### PR TITLE
[JAX] Fix GEMM partitioning to reduce over nested contracting mesh axes

### PR DESCRIPTION
**Pull request by phu0ngng**
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by xiaopoc** [comment]
Tested this PR on DeepSeek-V3 with 32 H100 GPUs using TP8, DP4, EP16, FSDP2, and ETP1. The previous global WGrad buffer/all-gather disappeared, HLO now shows WGrad reduce-scatter over the tensor axis, and the training loss decreases normally. This fixes the issue I observed, thanks.

**Review by xiaopoc — APPROVED** [review]
Looks good to me. Verified in the test described in my comment above.

File: tests/jax/test_gemm_partitioning.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+import jax
+from jax.sharding import Mesh, PartitionSpec
+
+from transformer_engine.jax.cpp_extensions.gemm import CollectiveOp, GemmPrimitive
+from transformer_engine.jax.quantize import ScalingMode
+from transformer_engine.jax.sharding import MeshResource, global_shard_guard
+
+
+def _mesh(axes):
+    """Build a CPU mesh with a size-2 device grid over the named axes."""
```
Comment (jberchtold-nvidia) [inline tests/jax/test_gemm_partitioning.py:47]: nit SR: GPU not CPU?

File: tests/jax/test_gemm_partitioning.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        ("fsdp", "expert"),
+    ),
+    # Mirror orientation: tp leaked onto dY's token dim only -> gather tp on dY.
+    "wgrad_mirror_tp_on_dy": Case(
+        ("fsdp", "tp", "expert"),
+        _MR_TP,
+        (7168, 524288),
+        (None, ("fsdp", "expert")),
+        (256, 524288),
+        ("tp", ("fsdp", "tp", "expert")),
```
Comment (jberchtold-nvidia) [inline tests/jax/test_gemm_partitioning.py]: Isn't specifying TP in the spec twice an invalid spec?

File: transformer_engine/jax/cpp_extensions/gemm.py
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/gemm.py:1005]: This is our WAR solution to handle other mesh axes participating in FSDP since our MeshResource does not allow us to specify contexts that persist thru the custom partitioning pass. For example, we do `expert`=EP in the MoEBlock, but `expert` acts as FSDP outside. 

So with a sharding of `((fsdp, expert), some_other_axis, None)` then we will AG `(fsdp, expert)` acting as FSDP, but not `some_other_axis`

I think this WAR is still correct with the current test cases with TP you've added above, but the one I'm not sure on is where rhs_spec = `("tp", ("fsdp", "tp", "expert"))`

File: transformer_engine/jax/cpp_extensions/gemm.py [hunk 3]
```diff
@@ -913,13 +913,31 @@ def _parse_operand_output_specs(
             (lhs_non_cdims, lhs_cdims, rhs_non_cdims, rhs_cdims),
         )
 
+        # A contracting-dim spec element can be a single mesh axis or a nested tuple of axes
+        # (e.g. ("fsdp", "tp", "expert")). Flatten so we can reason per mesh axis.
+        def _flatten_spec(spec):
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/gemm.py]: This is not really flattening, it's converting `(fsdp, tp, expert)` to `[fsdp, tp, expert]` and `tp` to `[tp]`

If a user input a nested tuple, e.g. the full tensor's partition spec `((fsdp, expert), tp, None)` this would return `[(fsdp, expert), tp, None]`

Should we rename to `_convert_axis_spec_to_list`?

Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/gemm.py]: Related, I think `spec` is an overloaded term since we have tuples that can be the full tensor's spec, but a tuple can also be an individual data axis's spec. Both `_flatten_spec` and `_retain_spec` only support the data axis spec as their `spec` argument. They are both used correctly in this PR so this is not a functionality blocker, I'm thinking for future changes and our codebase in general

Would the following naming split be helpful?
* `axis_spec`: spec of an individual data axis, could be a single mesh axis string or tuple of multiple mesh axes
* `tensor_spec`: spec of the full tensor, all data axes. Though this could be an overloaded name as well with tensor-parallelism

Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/gemm.py]: I think `_convert_axis_spec_to_list` would be helpful in this PR, we can discuss `spec` naming further offline and it doesn't need to block this PR. Thanks!

File: tests/jax/test_gemm_partitioning.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    "axes, resource, lhs_shape, lhs_spec, rhs_shape, rhs_spec, cdims,"
+    " exp_lhs, exp_rhs, exp_out, exp_reduce",
+)
+
+_MR_TP = dict(fsdp_resource="fsdp", tp_resource="tp", ep_resource="expert")
+_MR_TPSP = dict(fsdp_resource="fsdp", tpsp_resource="tp", ep_resource="expert")
+
+CASES = {
+    # WGrad, tp leaked onto X's token dim only -> gather tp on X, reduce (fsdp, expert).
+    "wgrad_nested_tp_on_x": Case(
```
Comment (jberchtold-nvidia) [inline tests/jax/test_gemm_partitioning.py]: I like these style of tests a lot! I will add similar tests for grouped GEMM when I add custom partitioning, in addition to our typical distributed tests that assert on expected numeric results.

These tests let us test a wide variety of parallelisms faster (relevant as we'll be essentially doubling our test suite when supporting explicit sharding)

And this could also catch issues like silent replication where numerical accuracy is unchanged but perf is impacted.

**Review by jberchtold-nvidia — COMMENTED** [review]
Overall looks good, thanks! Left a couple small comments and a question

File: tests/jax/test_gemm_partitioning.py (hunk 1 above)
Comment (phu0ngng) [reply tests/jax/test_gemm_partitioning.py:47]: We can run this test on the CPU since it only parses the infer sharding spec strings given input specs. I think it's also faster.

File: transformer_engine/jax/cpp_extensions/gemm.py (hunk 3 above)
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/gemm.py]: Renamed to `_convert_axis_spec_to_list` for now.

File: tests/jax/test_gemm_partitioning.py (hunk 2 above)
Comment (phu0ngng) [reply tests/jax/test_gemm_partitioning.py]: Double checking ...

Comment (phu0ngng) [reply tests/jax/test_gemm_partitioning.py]: Removed this invalid case. 
I also reworked the test suite so that it contains valid cases. 
The test is quite cheap; it takes 2.9s on my node.

Comment (jberchtold-nvidia) [reply tests/jax/test_gemm_partitioning.py]: That's great, super fast!

File: transformer_engine/jax/cpp_extensions/gemm.py
Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/gemm.py:1005]: Resolving as ("tp", ("fsdp", "tp", "expert")) case was invalid sharding spec

File: tests/jax/test_gemm_partitioning.py (hunk 1 above)
Comment (jberchtold-nvidia) [reply tests/jax/test_gemm_partitioning.py:47]: Oh good to know, nice!

**Review by jberchtold-nvidia — APPROVED** [review]
LGTM, thanks!
